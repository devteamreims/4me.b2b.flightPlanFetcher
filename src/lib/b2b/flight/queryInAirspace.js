import moment from 'moment';
import R from 'ramda';

import invariant from 'invariant';

import {
  b2bFormatDuration,
  b2bTimeFormatWithSeconds,
  b2bTimeFormat,
  flightSoapEnvelope,
  parseXML,
  unfoldSoapEnvelope,
  B2BError,
} from '../utils';

export function query(airspace = 'LFEERMS', options = {}) {
  invariant(
    typeof airspace === 'string',
    'Argument error: airspace should be a string',
  );

  invariant(
    typeof options === 'object',
    'Argument error: options should be an object',
  );

  const sendTime = moment.utc().format(b2bTimeFormatWithSeconds);

  const wef = moment.utc().subtract(10, 'minutes').format(b2bTimeFormat);
  const unt = moment.utc().add(20, 'minutes').format(b2bTimeFormat);

  const duration = b2bFormatDuration(R.propOr(11, 'duration', options));
  const step = b2bFormatDuration(R.propOr(1, 'step', options));

  const query = (`
     <sendTime>${sendTime}</sendTime>
     <dataset>
        <type>OPERATIONAL</type>
     </dataset>
     <includeProposalFlights>false</includeProposalFlights>
     <trafficType>LOAD</trafficType>
     <trafficWindow>
        <wef>${wef}</wef>
        <unt>${unt}</unt>
     </trafficWindow>
     <countsInterval>
        <duration>${duration}</duration>
        <step>${step}</step>
     </countsInterval>
     <!--Optional:-->
     <calculationType>OCCUPANCY</calculationType>
     <airspace>${airspace}</airspace>
  `);

  return flightSoapEnvelope(`
    <flig:FlightListByAirspaceRequest>
      ${query}
    </flig:FlightListByAirspaceRequest>
  `);
}

export function parseResponse(rawResponse) {
  const checkResponseStatus = b2bResponse => {
    const status = R.propOr('UNKNOWN', 'status', b2bResponse);
    if(status !== 'OK') {
      throw new B2BError(`Invalid response status : ${status}`, {b2bResponse});
    }

    return b2bResponse;
  };

  return R.pipeP(
    // Transform raw XML to JS
    parseXML,
    // Remove SOAP Envelope
    unfoldSoapEnvelope,
    // Extract specific response
    R.prop('flight:FlightListByAirspaceReply'),
    // Check B2B response status, throw if different than 'OK'
    checkResponseStatus,
  )(rawResponse);
}
