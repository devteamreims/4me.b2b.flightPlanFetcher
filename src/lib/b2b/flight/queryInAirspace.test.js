import { query, parseResponse } from './queryInAirspace';
import MockDate from 'mockdate';
describe('queryInAirspace', () => {
  describe('query builder', () => {
    // Our query build depends on the current time
    // MockDate allows snapshot testing by overriding global Date class
    beforeEach(() => {
      MockDate.set(1479898355004);
    });

    afterEach(() => {
      MockDate.reset();
    });

    test('should have proper defaults', () => {
      expect(query()).toMatchSnapshot();
    });

    test('should accept valid arguments', () => {
      expect(() => query()).not.toThrow();
      expect(() => query('string')).not.toThrow();
      expect(() => query('string', {})).not.toThrow();
    });

    test('should reject invalid arguments', () => {
      expect(() => query({})).toThrowErrorMatchingSnapshot();
      expect(() => query(null)).toThrowErrorMatchingSnapshot();
      expect(() => query('string', 'string')).toThrowErrorMatchingSnapshot();
    });

    test('should accept an airspace as argument', () => {
      expect(query('MOCKAIRSPACE')).toMatch(/MOCKAIRSPACE/);
    });

    test('should match snapshot', () => {
      expect(query('MOCKAIRSPACE')).toMatchSnapshot();
    });
  });

  describe('response parser', () => {
    const exampleResponse = `
  <S:Envelope xmlns:S="http://schemas.xmlsoap.org/soap/envelope/">
    <S:Body>
      <flight:FlightListByAirspaceReply xmlns:flight="eurocontrol/cfmu/b2b/FlightServices" xmlns:common="eurocontrol/cfmu/b2b/CommonServices" xmlns:airspace="eurocontrol/cfmu/b2b/AirspaceServices" xmlns:flow="eurocontrol/cfmu/b2b/FlowServices">
        <requestReceptionTime>2016-11-23 11:11:03</requestReceptionTime>
        <requestId>B2B_CUR:1884063</requestId>
        <sendTime>2016-11-23 11:11:03</sendTime>
        <status>OK</status>
        <data>
          <flights>
             <flight>
                <flightId>
                   <id>AA58786362</id>
                   <keys>
                      <aircraftId>EZY18AG</aircraftId>
                      <aerodromeOfDeparture>LSZH</aerodromeOfDeparture>
                      <nonICAOAerodromeOfDeparture>false</nonICAOAerodromeOfDeparture>
                      <airFiled>false</airFiled>
                      <aerodromeOfDestination>EGGW</aerodromeOfDestination>
                      <nonICAOAerodromeOfDestination>false</nonICAOAerodromeOfDestination>
                      <estimatedOffBlockTime>2016-11-23 09:35</estimatedOffBlockTime>
                   </keys>
                </flightId>
             </flight>
          </flights>
          <flights>
             <flight>
                <flightId>
                   <id>AA58788176</id>
                   <keys>
                      <aircraftId>BAW753</aircraftId>
                      <aerodromeOfDeparture>LFSB</aerodromeOfDeparture>
                      <nonICAOAerodromeOfDeparture>false</nonICAOAerodromeOfDeparture>
                      <airFiled>false</airFiled>
                      <aerodromeOfDestination>EGLL</aerodromeOfDestination>
                      <nonICAOAerodromeOfDestination>false</nonICAOAerodromeOfDestination>
                      <estimatedOffBlockTime>2016-11-23 10:10</estimatedOffBlockTime>
                   </keys>
                </flightId>
             </flight>
          </flights>
          <flights>
             <flight>
                <flightId>
                   <id>AA58793473</id>
                   <keys>
                      <aircraftId>UAL135</aircraftId>
                      <aerodromeOfDeparture>LSZH</aerodromeOfDeparture>
                      <nonICAOAerodromeOfDeparture>false</nonICAOAerodromeOfDeparture>
                      <airFiled>false</airFiled>
                      <aerodromeOfDestination>KEWR</aerodromeOfDestination>
                      <nonICAOAerodromeOfDestination>false</nonICAOAerodromeOfDestination>
                      <estimatedOffBlockTime>2016-11-23 09:50</estimatedOffBlockTime>
                   </keys>
                </flightId>
             </flight>
          </flights>
          <effectiveTrafficWindow>
            <wef>2016-11-23 10:42</wef>
            <unt>2016-11-23 11:22</unt>
          </effectiveTrafficWindow>
        </data>
      </flight:FlightListByAirspaceReply>
    </S:Body>
  </S:Envelope>
    `;

    const failedResponse = `
    <S:Envelope xmlns:S="http://schemas.xmlsoap.org/soap/envelope/">
       <S:Body>
          <flight:FlightListByAirspaceReply xmlns:flight="eurocontrol/cfmu/b2b/FlightServices" xmlns:common="eurocontrol/cfmu/b2b/CommonServices" xmlns:airspace="eurocontrol/cfmu/b2b/AirspaceServices" xmlns:flow="eurocontrol/cfmu/b2b/FlowServices">
             <requestReceptionTime>2016-11-23 11:57:11</requestReceptionTime>
             <requestId>B2B_CUR:1943185</requestId>
             <sendTime>2016-11-23 11:57:11</sendTime>
             <status>INVALID_INPUT</status>
             <inputValidationErrors>
                <attributes>
                   <item>countsInterval:step</item>
                </attributes>
                <group>FLIGHT</group>
                <category>GEN</category>
                <type>INVALID_ATTRIBUTE_VALUE</type>
                <parameters>
                   <item>
                      <key>RECEIVED</key>
                      <value>aze</value>
                   </item>
                   <item>
                      <key>CONSTRAINT</key>
                      <value>value does not respect the expected format: ' four digits 0 to 9'</value>
                   </item>
                </parameters>
             </inputValidationErrors>
          </flight:FlightListByAirspaceReply>
       </S:Body>
    </S:Envelope>
    `;

    test('should parse correct response', () => {
      return parseResponse(exampleResponse).then(
        parsedResponse => expect(parsedResponse).toMatchSnapshot(),
        () => expect(false).toBe(true),
      );
    });

    test('should handle failed response', () => {
      return parseResponse(failedResponse).then(
        () => expect(false).toBe(true),
        error => {
          expect(error.name).toBe('B2BError');
        }
      );
    });

  });
});
