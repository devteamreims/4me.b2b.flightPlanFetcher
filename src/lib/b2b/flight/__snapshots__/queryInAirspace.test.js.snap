exports[`queryInAirspace query builder should have proper defaults 1`] = `
"
    <soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\" xmlns:flig=\"eurocontrol/cfmu/b2b/FlightServices\">
      <soapenv:Header/>
      <soapenv:Body>
        
    <flig:FlightListByAirspaceRequest>
      
     <sendTime>2016-11-23 10:52:35</sendTime>
     <dataset>
        <type>OPERATIONAL</type>
     </dataset>
     <includeProposalFlights>false</includeProposalFlights>
     <trafficType>LOAD</trafficType>
     <trafficWindow>
        <wef>2016-11-23 10:42</wef>
        <unt>2016-11-23 11:12</unt>
     </trafficWindow>
     <countsInterval>
        <duration>0011</duration>
        <step>0001</step>
     </countsInterval>
     <!--Optional:-->
     <calculationType>OCCUPANCY</calculationType>
     <airspace>LFEERMS</airspace>
  
    </flig:FlightListByAirspaceRequest>
  
      </soapenv:Body>
    </soapenv:Envelope>
  "
`;

exports[`queryInAirspace query builder should match snapshot 1`] = `
"
    <soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\" xmlns:flig=\"eurocontrol/cfmu/b2b/FlightServices\">
      <soapenv:Header/>
      <soapenv:Body>
        
    <flig:FlightListByAirspaceRequest>
      
     <sendTime>2016-11-23 10:52:35</sendTime>
     <dataset>
        <type>OPERATIONAL</type>
     </dataset>
     <includeProposalFlights>false</includeProposalFlights>
     <trafficType>LOAD</trafficType>
     <trafficWindow>
        <wef>2016-11-23 10:42</wef>
        <unt>2016-11-23 11:12</unt>
     </trafficWindow>
     <countsInterval>
        <duration>0011</duration>
        <step>0001</step>
     </countsInterval>
     <!--Optional:-->
     <calculationType>OCCUPANCY</calculationType>
     <airspace>MOCKAIRSPACE</airspace>
  
    </flig:FlightListByAirspaceRequest>
  
      </soapenv:Body>
    </soapenv:Envelope>
  "
`;

exports[`queryInAirspace query builder should reject invalid arguments 1`] = `"Argument error: airspace should be a string"`;

exports[`queryInAirspace query builder should reject invalid arguments 2`] = `"Argument error: airspace should be a string"`;

exports[`queryInAirspace query builder should reject invalid arguments 3`] = `"Argument error: options should be an object"`;

exports[`queryInAirspace response parser should parse correct response 1`] = `
Object {
  "$": Object {
    "xmlns:airspace": "eurocontrol/cfmu/b2b/AirspaceServices",
    "xmlns:common": "eurocontrol/cfmu/b2b/CommonServices",
    "xmlns:flight": "eurocontrol/cfmu/b2b/FlightServices",
    "xmlns:flow": "eurocontrol/cfmu/b2b/FlowServices",
  },
  "data": Object {
    "effectiveTrafficWindow": Object {
      "unt": "2016-11-23 11:22",
      "wef": "2016-11-23 10:42",
    },
    "flights": Array [
      Object {
        "flight": Object {
          "flightId": Object {
            "id": "AA58786362",
            "keys": Object {
              "aerodromeOfDeparture": "LSZH",
              "aerodromeOfDestination": "EGGW",
              "airFiled": "false",
              "aircraftId": "EZY18AG",
              "estimatedOffBlockTime": "2016-11-23 09:35",
              "nonICAOAerodromeOfDeparture": "false",
              "nonICAOAerodromeOfDestination": "false",
            },
          },
        },
      },
      Object {
        "flight": Object {
          "flightId": Object {
            "id": "AA58788176",
            "keys": Object {
              "aerodromeOfDeparture": "LFSB",
              "aerodromeOfDestination": "EGLL",
              "airFiled": "false",
              "aircraftId": "BAW753",
              "estimatedOffBlockTime": "2016-11-23 10:10",
              "nonICAOAerodromeOfDeparture": "false",
              "nonICAOAerodromeOfDestination": "false",
            },
          },
        },
      },
      Object {
        "flight": Object {
          "flightId": Object {
            "id": "AA58793473",
            "keys": Object {
              "aerodromeOfDeparture": "LSZH",
              "aerodromeOfDestination": "KEWR",
              "airFiled": "false",
              "aircraftId": "UAL135",
              "estimatedOffBlockTime": "2016-11-23 09:50",
              "nonICAOAerodromeOfDeparture": "false",
              "nonICAOAerodromeOfDestination": "false",
            },
          },
        },
      },
    ],
  },
  "requestId": "B2B_CUR:1884063",
  "requestReceptionTime": "2016-11-23 11:11:03",
  "sendTime": "2016-11-23 11:11:03",
  "status": "OK",
}
`;
